manticore/core/smtlib/solver.py

```diff
@@ -735,18 +770,162 @@ def __init__(self):
 
 class CVC4Solver(SMTLIBSolver):
     def __init__(self):
-        init = ["(set-logic QF_AUFBV)", "(set-option :produce-models true)"]
-        command = f"{consts.cvc4_bin} --tlimit={consts.timeout * 1000} --lang=smt2 --incremental"
+        init = info.inits["cvc4"]
+        command = info.commands["cvc4"]
+        self.ncores = 1
         super().__init__(command=command, init=init)
 
 
 class BoolectorSolver(SMTLIBSolver):
-    def __init__(self):
-        init = ["(set-logic QF_AUFBV)", "(set-option :produce-models true)"]
-        command = f"{consts.boolector_bin} --time={consts.timeout} -i"
+    def __init__(self, args: List[str] = []):
+        init = info.inits["boolector"]
+        command = info.commands["boolector"]
+        self.ncores = 1
         super().__init__(command=command, init=init)
 
 
+class SmtlibPortfolio:
+    def __init__(self, solvers: List[str], debug: bool = False):
+        """Single smtlib interactive process
+
+        :param command: the shell command to execute
+        :param debug: log all messaging
+        """
+        self._procs: List[SmtlibProc] = []
+        self._solvers = solvers
+        self._debug = debug
+        self._check_sat = False
+
+    def start(self):
+        if len(self._procs) == 0:
+            for solver in self._solvers:
+                self._procs.append(SmtlibProc(info.commands[solver], self._debug))
+
+        for proc in self._procs:
+            proc.start()
+
+    def stop(self):
+        """
+        Stops the solver process by:
+        - sending a SIGKILL signal,
+        - waiting till the process terminates (so we don't leave a zombie process)
+        """
+        for proc in self._procs:
+            proc.stop()
+
+        self._check_sat = False
+
+    def send(self, cmd: str) -> None:
+        """
+        Send a string to the solver.
+
+        :param cmd: a SMTLIBv2 command (ex. (check-sat))
+        """
+        assert len(self._procs) > 0
+        # print(cmd)
+        for proc in self._procs:
+            if not proc.is_started():
+                continue
+
+            proc.send(cmd)
+
+        if "(check-sat)" in cmd:
```

Comment:
What does _check_sat means?
Does this class accept a single check-sat command?

Reply:
In certain functions like `get_all_values` you check for sat and then ask for some values (with some more constrains):

```python
 while self._is_sat():
    value = self._getvalue(var)
   ...
```

 When you call `check-sat`, you start all the solvers. Eventually one of them will answer.  However, the other ones are going to be busy so when you start adding more constrains, so you need to discard the other solvers until you finish with the call and then restart the solvers. In summary, if `(check-sat)` was used, the portfolio solver should discard the executions of the remaining ones.

Reply:
Does that means that after the first `(check-sat)` all commands are in race mode?

Maybe then `_check_sat` should be reset to false after the recv?

Reply:
Before check-sat all solvers are in race mode. After, I should find a good way to resend all the smtlib2 commands so far.

Reply:
All the other solvers should be stopped after checking for sat, because we cannot use them in parallel with the use of `get-value`. The possibility of interleaving, while small, could result in invalid results (where some values are obtained from one solver and others are obtained from another one).